checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -1625,6 +1653,146 @@ && varTrackedInObligations(obligations, (LocalVariableNode) receiver))
+   * @param currentAssignment the actual assignment tree being analyzed, which is a statement in
+   *     {@code constructor}
+   * @return true if this assignment can be safely considered the first write during construction
+   */
+  private boolean isFirstAssignmentToField(
+      VariableElement field, MethodTree constructor, @FindDistinct Tree currentAssignment) {
+    TreePath constructorPath = cmAtf.getPath(constructor);
+    ClassTree classTree = TreePathUtil.enclosingClass(constructorPath);
+    String fieldName = field.getSimpleName().toString();
+
+    // (2) Disallow non-null inline initializer
+    for (Tree member : classTree.getMembers()) {
+      if (member instanceof VariableTree) {
+        VariableTree decl = (VariableTree) member;
+        if (decl.getName().contentEquals(fieldName) && decl.getInitializer() != null) {
+          if (decl.getInitializer().getKind() != Tree.Kind.NULL_LITERAL) {
+            return false;
+          }
+        }
+      }
+    }
+
+    // (3) Disallow assignment in instance initializer blocks
+    for (Tree member : classTree.getMembers()) {
+      if (member instanceof BlockTree) {
+        BlockTree block = (BlockTree) member;
+        if (block.isStatic()) {
+          continue;
+        }
+        // The variables accessed from within the inner class need to be effectively final, so
+        // AtomicBoolean is used here.
+        AtomicBoolean found = new AtomicBoolean(false);
+        block.accept(
+            new TreeScanner<Void, Void>() {
+              @Override
+              public Void visitAssignment(AssignmentTree node, Void unused) {
+                ExpressionTree lhs = node.getVariable();
+                Name lhsName =
+                    lhs instanceof MemberSelectTree
+                        ? ((MemberSelectTree) lhs).getIdentifier()
+                        : lhs instanceof IdentifierTree ? ((IdentifierTree) lhs).getName() : null;
+                if (lhsName != null && lhsName.contentEquals(fieldName)) {
+                  found.set(true);
+                }
+                return super.visitAssignment(node, unused);
+              }
+            },
+            null);
+        if (found.get()) {
+          return false;
+        }
+      }
+    }
+
+    // (4) Check constructor initialization chaining
+    if (callsThisConstructor(constructor)) {
+      return false;
+    }
+
+    // (5) Check for earlier assignments to the same field, or
+    // (6) any method calls other than super.
+    List<? extends StatementTree> statements = constructor.getBody().getStatements();
+    for (StatementTree stmt : statements) {
+      if (!(stmt instanceof ExpressionStatementTree)) {
+        continue;
+      }
+      ExpressionTree expr = ((ExpressionStatementTree) stmt).getExpression();
+      // Stop when we reach the current assignment
+      if (expr == currentAssignment) {
+        break;
+      }
+      // (5) Prior assignment to the same field
```

Comment:
I think this (or possibly some other bit of similar code) causes errors to be issued at the wrong location.  See added test, which fails.